File: evalboard/lib/timing.ts
import type { MessageEvent } from "./runs";

// Wall-clock efficiency: how a task's duration compares to the time it is
// expected to take.
//
// `expected_seconds` is derived per task, per harness by the eval runner (p10 of
// past successful durations) and stamped into run.json. The dashboard reads that
// stamp rather than deriving its own, so a number here always matches the line the
// run was actually scored against.
//
// A task with no `expected_seconds` is *unscored*, not "within budget": too young
// for history, or a run that predates stamping. Every helper returns null there.

export interface TimeRatioThresholds {
    yellow: number;
    red: number;
}

export function getTimeRatioThresholds(): TimeRatioThresholds {
    const parse = (raw: string | undefined, fallback: number) => {
        if (!raw) return fallback;
        const n = Number(raw);
        return Number.isFinite(n) && n > 0 ? n : fallback;
    };
    return {
        yellow: parse(process.env.EVALBOARD_TIME_YELLOW_RATIO, 1.5),
        red: parse(process.env.EVALBOARD_TIME_RED_RATIO, 2),
    };
}

export type TimeTint = "green" | "yellow" | "red" | null;

// Pure time-efficiency ratio (seconds ÷ expected_seconds), used to tint per-task
// duration cells. Blind to pass/fail on purpose: the cell answers "did this take
// longer than it should?", which holds either way. The aggregates diverge
// deliberately and score passes only (withinExpectedTime, overview.ts).
export function timeRatio(
    durationSeconds: number | null,
    expectedSeconds: number | null,
): number | null {
    if (
        durationSeconds == null ||
        expectedSeconds == null ||
        expectedSeconds <= 0
    ) {
        return null;
    }
    return durationSeconds / expectedSeconds;
}

export function tintForTimeRatio(
    ratio: number | null,
    t: TimeRatioThresholds = getTimeRatioThresholds(),
): TimeTint {
    if (ratio == null) return null;
    if (ratio > t.red) return "red";
    if (ratio > t.yellow) return "yellow";
    return "green";
}

export function timeCellClasses(tint: TimeTint): string {
    switch (tint) {
        case "green":
            return "text-emerald-700";
        case "yellow":
            return "text-amber-700";
        case "red":
            return "text-rose-700";
        default:
            return "text-gray-900";
    }
}

// A task counts as within its expected time while it stays inside
// (1 + tolerance) × expected, so anything past 2× its line reads slow. Mirrors
// `timing.TOLERANCE` on the runner side, which records the value it used in each
// run's `timing` block. Red tints at the same 2×, so the cell and the rollup agree.
export const TIME_BUDGET_TOLERANCE = 1;

// Whether a task came in at or under (1 + tolerance) × its expected time.
// Null when the task is not scoreable: no duration, or no positive
// `expected_seconds`.
export function withinExpectedTime(
    durationSeconds: number | null,
    expectedSeconds: number | null,
    tolerance: number = TIME_BUDGET_TOLERANCE,
): boolean | null {
    const ratio = timeRatio(durationSeconds, expectedSeconds);
    if (ratio == null) return null;
    return ratio <= 1 + tolerance;
}

// Per-task wall clock, to the second: `3m14s`, `1h02m` past the hour. Seconds
// are the point of the metric, so they are never rounded away below an hour.
export function fmtTaskSeconds(seconds: number | null): string {
    if (seconds == null) return "—";
    const s = Math.round(seconds);
    const h = Math.floor(s / 3600);
    const m = Math.floor((s % 3600) / 60);
    const sec = s % 60;
    return h
        ? `${h}h${String(m).padStart(2, "0")}m`
        : `${m}m${String(sec).padStart(2, "0")}s`;
}

export function fmtTimeRatio(ratio: number | null): string {
    return ratio == null ? "—" : `${ratio.toFixed(2)}x expected`;
}

// Ratio as a table cell: `1.8×`, or an em dash for an unscored task. One decimal,
// not two: the baseline is a min over a handful of runs (p10 over ten), and a
// task's own night-to-night spread is wider than the digit a second decimal adds.
export function fmtTimeRatioCell(ratio: number | null): string {
    return ratio == null ? "—" : `${ratio.toFixed(1)}×`;
}

// Title text for a duration cell: what the task was measured against, or why it
// was not measured at all.
export function expectedTimeTitle(expectedSeconds: number | null): string {
    return expectedSeconds != null
        ? `expected time: ${fmtTaskSeconds(expectedSeconds)}`
        : "no expected time yet (needs a passing run on this harness)";
}

// Epoch milliseconds for a `CommandTelemetry` timestamp, or null when it is
// absent or unparseable. The stamps are naive local ISO strings (Python
// `datetime.now()`), which `Date` reads as local time — every stamp in a run
// comes from one machine, and only DIFFERENCES are ever taken, so the offset
// cancels.
export function epochMs(value: string | null | undefined): number | null {
    if (typeof value !== "string" || !value) return null;
    const t = Date.parse(value);
    return Number.isFinite(t) ? t : null;
}

// Milliseconds inside [lo, hi] where at least ONE span was running: the UNION,
// not the sum.
//
// The TypeScript twin of `coder_eval.timing.busy_ms`, deliberately the
// same algorithm — the harness subtracts tool time from its generation windows
// with it (once, in coder_eval/timing.py::subtract_tool_time), and this file
// subtracts tool time from a task's wall clock, so the two must agree about
// what "tool execution took N ms" means. Held in step by
// tests/_fixtures/timing_union_cases.json, which both suites replay.
export function busyMs(
    spans: [number, number][],
    lo: number,
    hi: number,
): number {
    const clipped = spans
        .map(([s, e]) => [Math.max(s, lo), Math.min(e, hi)] as [number, number])
        .filter(([s, e]) => e > s)
        .sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    if (clipped.length === 0) return 0;
    let total = 0;
    let [openStart, openEnd] = clipped[0];
    for (const [start, end] of clipped.slice(1)) {
        if (start > openEnd) {
            total += openEnd - openStart;
            [openStart, openEnd] = [start, end];
        } else {
            openEnd = Math.max(openEnd, end);
        }
    }
    return total + (openEnd - openStart);
}

// Wall-clock milliseconds these messages' tool calls occupied: the UNION of
// their BOUNDED execution intervals.
//
// Concurrent tools occupy the wall clock once, and summing them drove the task
// page's Unaccounted cell to -615ms on a task with two concurrent sleeps, where
// the honest answer was +2.5s of sandbox setup and grading.
//
// A call the harness TIMED but did not BOUND contributes nothing — no
// `durationMs` fallback. That is a policy, and it is the same one
// `coder_eval.timing.main_thread_tool_spans` has always had on the Python side:
// its `is not None` filter drops a command with no `execution_started_at` /
// `execution_completed_at`, so every Python surface already ignored these while
// this function folded them in. A duration with no bounds cannot be placed on
// the timeline, so it cannot be unioned with anything; adding it to a union
// double-books whatever it overlapped and can drive the residual negative,
// which destroys the disjointness the four-bucket identity rests on. Such time
// lands in Unaccounted instead, which is precisely what that cell is for — the
// harness measured a duration it cannot place.
//
// Measured blast radius: 9336 of 12170 commands in the run history on disk are
// unbounded, every one a codex `Bash`, ~8 h in aggregate. That population is
// CLOSED — codex went from 0% bounded before 2026-09-10 to 100% after — so no
// future run changes, but on historical codex runs this moves up to ~8 h out of
// Tool exec and into Unaccounted. Going forward the only harness reporting a
// bare duration is the out-of-tree `delegate-sdk`; see
// docs/agents/HARNESS_PARITY.md.
export function toolExecutionMs(messages: MessageEvent[]): number {
    const spans: [number, number][] = [];
    for (const m of messages) {
        for (const t of m.toolUses) {
            if (t.execStartMs != null && t.execEndMs != null) {
                spans.push([t.execStartMs, t.execEndMs]);
            }
        }
    }
    if (spans.length === 0) return 0;
    const lo = Math.min(...spans.map(([s]) => s));
    const hi = Math.max(...spans.map(([, e]) => e));
    return busyMs(spans, lo, hi);
}

File: src/coder_eval/timing.py
"""Wall-clock arithmetic for a turn, defined once and shared.

A cycle-free leaf: it sits outside ``agents/`` because ``EventCollector``
consumes it, and importing anything under ``agents/`` pulls in every agent,
which imports ``streaming/``.

NO harness subtracts tool execution from its own generation windows. Each
publishes the RAW window it measured, and ``subtract_tool_time`` takes the
UNION of the tool intervals back out of them once, for all five, at the single
capture seam. A reducer's only remaining timing decision is where its window
opens.

There is a TypeScript twin, ``evalboard/lib/timing.ts::busyMs``, answering the
same question about the same ``task.json``, so the two must agree — and neither
owns the numbers: ``tests/_fixtures/timing_union_cases.json`` does, and both
suites replay it.

Rationale: .claude/notes/timing.md § Where a reducer's window opens
"""

import math
import time
from collections.abc import Iterable
from datetime import datetime, timedelta

from coder_eval.models import AssistantMessage, CommandTelemetry, TranscriptMessage


class TurnClock:
    """One (wall, monotonic) pair per turn; every later stamp derives from it.

    A turn's bounds and its durations must share a basis, or they disagree in a
    field measured in milliseconds. Stamps stay NAIVE LOCAL, matching the rest of
    the telemetry and the persisted ``execution_started_at``.

    ONE PER TURN, never module-level and never reused across turns: a long run
    accumulates drift between the pair and real wall time. Within a turn the
    derived stamp is monotonic-accurate and may drift from real wall time; each
    turn re-anchors. That is intended — do not "fix" it by re-reading the wall
    clock, which is the property being removed.

    NOT for deadlines. Those stay on ``time.monotonic()`` directly: a deadline
    must not move when the wall clock steps.

    Which harnesses use it is the `clock basis for recorded stamps` row in
    docs/agents/HARNESS_PARITY.md, the SSOT for per-harness composition.

    Rationale: .claude/notes/timing.md § TurnClock
    """

    def __init__(self) -> None:
        self._wall0 = datetime.now()
        self._mono0 = time.monotonic()

    def now(self) -> datetime:
        return self._wall0 + timedelta(seconds=time.monotonic() - self._mono0)


def _require_same_awareness(a: datetime, b: datetime, *, field: str) -> None:
    """Raise if one stamp is timezone-aware and the other is naive.

    Only the MIX raises: an agent consistently aware, or consistently naive, is
    not this function's problem. It turns a bare ``TypeError`` raised deep in the
    arithmetic into a statement of which pair disagreed and which side is aware.

    Rationale: .claude/notes/timing.md § _require_same_awareness
    """
    if (a.tzinfo is None) == (b.tzinfo is None):
        return
    aware, naive = ("first", "second") if a.tzinfo is not None else ("second", "first")
    raise TypeError(
        f"{field}: one stamp is timezone-aware and the other is naive (the {aware} is aware, "
        + f"the {naive} is not), so the interval between them cannot be measured. Every stamp "
        + "this harness records is a naive local `datetime.now()`; if you are writing an agent "
        + "outside this repo (the `coder_eval.plugins` SPI), make its stamps naive local too "
        + "rather than normalizing here, so its tool spans and its window bounds keep one basis."
    )


def busy_ms(spans: list[tuple[datetime, datetime]], lo: datetime, hi: datetime) -> float:
    """Wall milliseconds inside ``[lo, hi]`` where at least ONE span was running.

    The UNION, not the sum: overlapping tool intervals would otherwise over-count
    the busy time by exactly the overlap and drive a generation window negative.
    Spans are clipped to ``[lo, hi]``, so a tool that opened earlier contributes
    only the part that ran inside this window.

    Raises ``TypeError`` (via ``_require_same_awareness``) if the bounds and the
    spans do not share a timezone awareness. An EMPTY span list returns ``0.0``
    and is not checked at all.

    Rationale: .claude/notes/timing.md § busy_ms
    """
    if not spans:
        return 0.0
    _require_same_awareness(lo, hi, field="busy_ms window")
    for span_start, span_end in spans:
        _require_same_awareness(lo, span_start, field="busy_ms window vs a tool span's start")
        _require_same_awareness(hi, span_end, field="busy_ms window vs a tool span's end")
    clipped = sorted((max(s, lo), min(e, hi)) for s, e in spans if min(e, hi) > max(s, lo))
    if not clipped:
        return 0.0
    total = 0.0
    open_start, open_end = clipped[0]
    for start, end in clipped[1:]:
        if start > open_end:  # disjoint — bank the run and start a new one
            total += (open_end - open_start).total_seconds() * 1000.0
            open_start, open_end = start, end
        else:  # overlapping or adjacent — extend the run
            open_end = max(open_end, end)
    return total + (open_end - open_start).total_seconds() * 1000.0


def union_ms(spans: list[tuple[datetime, datetime]]) -> float:
    """Wall milliseconds at least ONE span was running, over their full extent.

    ``busy_ms`` with the window set to the spans' own bounds.

    It does NOT filter ``end < start``. EVERY caller drops those while building
    its span list, so a new caller that skips the check gets ``busy_ms``'s
    silent discard rather than this function's stated contract.

    Rationale: .claude/notes/timing.md § union_ms
    """
    if not spans:
        return 0.0
    return busy_ms(spans, min(s for s, _ in spans), max(e for _, e in spans))


def close_window(*, mark: datetime, now: datetime, item_start: datetime | None = None) -> tuple[datetime, float]:
    """Open one generation window at ``mark`` and close it at ``now``: its ``(started, span_ms)``.

    The shape all five reducers share, and what it returns is the RAW window —
    tool execution comes back out centrally, in ``subtract_tool_time``.

    ``mark`` is where the window opens: the previous flush's close, which is what
    makes the windows TILE the turn contiguously. It is keyword-only with NO
    default so that no reducer can open a window without stating what it tiles
    from.

    ``item_start`` is this emission's own first stamp, when the harness has one;
    the ``min()`` against ``mark`` stops a backwards stamp inverting the span.

    The span is clamped at ``0.0``: an inverted window is a measured zero, not a
    negative generation. ``completed`` is deliberately not returned — it is
    always ``now``, which the caller already has.

    Rationale: .claude/notes/timing.md § close_window
    """
    started = min(mark, item_start) if item_start is not None else mark
    return started, max(0.0, (now - started).total_seconds() * 1000.0)


def decompose_turn(
    first_started_at: datetime | None,
    last_completed_at: datetime | None,
    agent_started_at: datetime | None,
    agent_ended_at: datetime | None,
    tool_spans: list[tuple[datetime, datetime]] | None = None,
) -> tuple[float | None, float | None]:
    """Wall ms before the first generation window opens, and after the last closes.

    The turn's two unexplained ends. Between them the windows tile and tool
    execution is already subtracted, so head + generation + UNION(tool) + tail
    is the whole turn.

    ``tool_spans`` keeps the four buckets DISJOINT: a tool is not confined to a
    generation window, so a span left in the head or tail is booked twice.

    ``EventCollector`` is the SOLE caller: computed once, persisted on
    ``TurnRecord``, READ by everyone later.

    The head means ONE thing on all five harnesses: wall clock from the turn
    start until the harness first observed model output. What it CONTAINS
    differs per harness (docs/agents/HARNESS_PARITY.md).

    ``None`` means never measured — never ``0.0`` (CE058). A measured inversion
    IS a real zero and clamps, because both ends were observed.

    Raises ``TypeError`` (via ``_require_same_awareness``) on a mixed pair.

    Rationale: .claude/notes/timing.md § decompose_turn
    """
    spans = tool_spans or []
    head = tail = None
    if first_started_at is not None and agent_started_at is not None:
        _require_same_awareness(agent_started_at, first_started_at, field="harness_startup_ms")
        elapsed = (first_started_at - agent_started_at).total_seconds() * 1000.0
        head = max(elapsed - busy_ms(spans, agent_started_at, first_started_at), 0.0)
    if last_completed_at is not None and agent_ended_at is not None:
        _require_same_awareness(last_completed_at, agent_ended_at, field="harness_teardown_ms")
        elapsed = (agent_ended_at - last_completed_at).total_seconds() * 1000.0
        tail = max(elapsed - busy_ms(spans, last_completed_at, agent_ended_at), 0.0)
    return head, tail


def main_thread_tool_spans(
    messages: Iterable[TranscriptMessage], commands: Iterable[CommandTelemetry]
) -> list[tuple[datetime, datetime]]:
    """Bounded execution intervals of the MAIN THREAD's tool calls.

    The span set the generation subtraction, the head and the tail are all
    measured against, so they cannot disagree about which calls exist. Shared
    with ``result_metrics.turn_time_buckets``;
    ``tests/test_timing_close_window.py::TestTheThreeToolUnionsAgree`` pins this,
    that, and ``scripts/timing/decompose_run.py`` together.

    Sub-agent tools are excluded — a child's calls are already covered by the
    parent Agent call's own interval. A sub-agent's tool ids are reachable only
    through the messages that own them: a child generation carries
    ``parent_tool_use_id``, and its ``tool_use_ids`` are the calls it made.

    An inverted pair (``end`` before ``start``) is dropped here, which is what
    keeps ``union_ms``'s "every caller filters" contract true.

    Rationale: .claude/notes/timing.md § main_thread_tool_spans
    """
    sub_agent_tool_ids = {
        tool_id
        for m in messages
        if isinstance(m, AssistantMessage) and m.parent_tool_use_id is not None
        for tool_id in m.tool_use_ids
    }
    return [
        (c.execution_started_at, c.execution_completed_at)
        for c in commands
        if c.execution_started_at is not None
        and c.execution_completed_at is not None
        and c.execution_completed_at >= c.execution_started_at
        and c.tool_id not in sub_agent_tool_ids
    ]


#: How far a published window may sit from the span its own bounds describe.
#: ONE MILLISECOND — the coarsest unit a field named ``_ms`` can honestly be
#: published in, so a producer that rounds is admitted while the defect class
#: (a reducer narrowing or widening a window without moving its bounds) is tens
#: to thousands of ms, orders of magnitude above it.
#: Rationale: .claude/notes/timing.md § _WINDOW_TOLERANCE_MS
_WINDOW_TOLERANCE_MS = 1.0


def subtract_tool_time(
    messages: list[TranscriptMessage],
    spans: list[tuple[datetime, datetime]],
) -> list[TranscriptMessage]:
    """Take tool execution back out of the generation windows it overlapped.

    THE one place this happens. Reducers publish RAW windows and never subtract.

    Returns a NEW list; messages are copied, never mutated — agents alias their
    own live message objects into ``AgentEndEvent``, so writing in place would
    reach back into agent state.

    Windows group by identical ``(started_at, completed_at)``, so the two
    sub-messages Codex splits one window into share a single subtraction.
    Sub-agent generations (``parent_tool_use_id`` set) are skipped. A
    ``generation_duration_ms`` of ``None`` passes through untouched — never
    coerced to ``0.0`` (CE058) — as does every non-``AssistantMessage`` entry.
    A window fully covered by tool execution reaches ``0.0``: a measurement.

    Raises ``ValueError`` if a group's published total does not equal the span
    its bounds describe. That equality is what lets ``generation_duration_ms``
    stay a PUBLISHED field; CE061 forces the same shape statically. Raising
    kills the turn, which is accepted.

    Rationale: .claude/notes/timing.md § subtract_tool_time
    """
    # (index, raw window ms) per group, captured while the message is already
    # narrowed to AssistantMessage.
    groups: dict[tuple[datetime, datetime], list[tuple[int, float]]] = {}
    for index, message in enumerate(messages):
        if not isinstance(message, AssistantMessage):
            continue
        raw = message.generation_duration_ms
        if raw is None or message.parent_tool_use_id is not None:
            continue
        groups.setdefault((message.started_at, message.completed_at), []).append((index, raw))

    out = list(messages)
    for (started, completed), members in groups.items():
        raw_total = sum(raw for _, raw in members)
        # Nothing to apportion, and the loop below divides by it — a group already
        # at zero stays at zero.
        #
        # ORDER IS LOAD-BEARING: this skip runs BEFORE the equality check below,
        # because `close_window` clamps an inverted window to 0.0 while its bounds
        # still say `completed_at < started_at` — a measured inversion the check
        # would otherwise raise on.
        # Rationale: .claude/notes/timing.md § subtract_tool_time
        if raw_total <= 0:
            continue
        bounds_ms = (completed - started).total_seconds() * 1000.0
        if not math.isclose(raw_total, bounds_ms, rel_tol=1e-9, abs_tol=_WINDOW_TOLERANCE_MS):
            raise ValueError(
                f"generation_duration_ms: a group of {len(members)} message(s) bounded "
                + f"{started} -> {completed} ({bounds_ms:.6f} ms) publishes {raw_total:.6f} ms of "
                + "generation. A reducer publishes the RAW window it measured, so its duration is "
                + "`completed_at - started_at` (or, across the sub-messages Codex splits one window "
                + "into, sums to it) — tool execution comes back out HERE, once, for every harness. "
                + "A disagreement means the reducer narrowed or widened a window without moving its "
                + "bounds, which makes the duration and the bounds two answers to one question and "
                + "breaks the four-bucket identity. Build the window with `timing.close_window` and "
                + "write `completed_at=now` (CE061), rather than adjusting the duration in place."
            )
        net = max(raw_total - busy_ms(spans, started, completed), 0.0)
        assigned = 0.0
        for n, (index, raw) in enumerate(members):
            # The last member takes the remainder so the parts reconstruct the
            # group's net exactly. NOT rounded: rounding an earlier share up could
            # push `assigned` past `net` and hand the last member a NEGATIVE
            # duration.
            share = net - assigned if n == len(members) - 1 else net * (raw / raw_total)
            out[index] = out[index].model_copy(update={"generation_duration_ms": share})
            assigned += share
    return out
